### PR TITLE
Add new options to the checksum checker drush cmd

### PR DESCRIPTION
**Pull request by jonathangreen**
**JIRA Ticket**: (link)
https://jira.duraspace.org/browse/ISLANDORA-1915

# What does this Pull Request do?

This pull request attempts to help answer the question: how many checksums should the checksum checker check when the checksum checker is checking checksums.

The checksum checker options ask the user to specify the number of items per invocation of the batch process. This is difficult to estimate if your repository is continually changing in size. Because generally you want to check all your checksums in a certain amount of time.

This pull request adds extra optional arguements to the checksum checker drush command that let the checksum checker determine the limit based on the number of items in your repository and how long you want it to take to check all your checksums.

# What's new?

This pull request adds two optional arguments to the drush command for the checksum checker. If both of these arguments are specified then the drush command will calculate the number of items that it processes per iteration in order to meet its time target based on the size of the repository. 

If these two arguments are not given then everything should work as it did before.

# How should this be tested?

Run the checksum checker as before with drush and no arguments. Make sure that it functions as it did before. Run the checksum checker with new arguments specified, ensure that the correct number of objects are processed based on the data passed in the arguments.

# Interested parties
@mjordan @DiegoPino

### CODE REVIEW

**Comment by mjordan** [comment]
@jonathangreen I will check this (but probably not until end of day tomorrow). BTW, one Travis check failed. Got time to check that out? :wink:

**Comment by jonathangreen** [comment]
@mjordan Updated! Thanks.

**Comment by mjordan** [comment]
I'm testing the new options by running the following in my staging server's crontab:

`0,30 * * * * /usr/bin/drush --root=/var/www/html --user=admin --days-to-complete=2 --cmd-run-frequency=1 run-islandora-checksum-queue`

So you may not hear back for up to two days....

**Comment by mjordan** [comment]
@jonathangreen this appears to work as intended, but I'd like to run it for another day or two to see the effect of changing the new options.

I also have a suggestion. I understand why the user needs to indicate `--cmd-run-frequency`, but I think it should be documented that this value corresponds to how cron should be configured to run the command. In other words, if you indicate a `--cmd-run-frequency=1`, you should also configure cron to run the drush command every hour. Otherwise the predicated number of objects per cron run will be inaccurate, and it will take a different number of days to check every object than the number estimated by the new options. Does this make sense? Maybe documenting this at the end of the "Frequency of verification" section? I'd be happy to provide some wording if you want.

**Comment by DiegoPino** [comment]
@mjordan @jonathangreen I agree on documenting that option as good as possible. I'm sure that is a task our release documenters could need help with (from you both). We are still in time to get this into the release (no panic! 😵 ). Let's wait then for your tests to finish. Thanks!

**Comment by jonathangreen** [comment]
I totally agree @mjordan that more documentation is good. I was struggling a bit to come up with a good description, so if you have any suggestions for wording that would be great.

**Comment by mjordan** [comment]
Changing the new options' values results in predictable values for the number of objects checked per cron run, so I'm giving this the :+1: and it can be merged, pending README updates.

But, one last comment: minimum allowed value of `--cmd-run-frequency` is 1. What if you run checksum checker more frequently, like every 30 or 15 minutes? On our staging server, where I've been doing my testing, average time to check each object is .5 seconds (most objects are not large). On large production sites with lots of video, it might take an average of 20 seconds to check each object. So for largish production sites, an hour might be a reasonable minimum run frequency.

I had totally forgotten about some work I had done back in September to try to gather some data to help admins calculate what @jonathangreen's new options accomplish. It's at https://github.com/mjordan/miscellaneous_islandora_docs/blob/master/calculating_checksum_checker_jobs.md. The most useful outcome of my tinkering was that "the length of time it takes to validate a checksum is proportionate to the size of the datastream." Perhaps we can include in the updated README that admins might want to keep this in mind when providing values for the new options?

**Comment by DiegoPino** [comment]
@mjordan thanks! Please add your review then and feel free to merge. Questions: Pending README updates means, in this pull, or as a secondary /documentation task? If so(secondary), @jonathangreen could you please add a documentation Jira ticket (i know, it's an overkill, but seems to be the only way we can keep track of this). Thanks a lot to both of you on your hard work on this.

**Comment by mjordan** [comment]
@DiegoPino I was thinking that the updates to README could be included in this PR since they are just documenting the new options. I am drafting some wording as per @jonathangreen's request so that he can decide if he wants to include something based on what I am providing. Is that OK?

But of course it's your call.

**Comment by DiegoPino** [comment]
@mjordan I'm fully Ok and very happy with this. Mostly trying to be a facilitator but did not wanted to push you into impossible deadlines, it's your call, you all know which wording works better 👍 thanks!

**Comment by mjordan** [comment]
Here's some text. I'd add it to the end of the "Frequency of verification" section but I'm fine with it going elsewhere if others see fit:

Checksum Checker provides two drush options that will calculate the number of objects to check per cron run. They are

* `--days-to-complete`: The number of days it should take to verify the checksums of  all objects.	
* `--cmd-run-frequency`: The number of hours between checksum runs (required if `--days-to-complete` is set). Note that this value (`1` for each hour, `2` for every 2 hours, etc.) should correspond to the frequency configured in your server's crontab for running the drush command.

Using these two values, and the number of objects in your repository (which you do not need to provide), Checksum Checker will calculate the number of objects to check each cron run.

**Comment by DiegoPino** [comment]
@mjordan I have never used this (shame on me!), but does `--cmd-run-frequency` takes on account if a previous "run" finished? or this is just about timing/defining when to start/how long to run?

**Comment by mjordan** [comment]
@DiegoPino each cron run adds objects to the end a Drupal queue, and then processes objects at the start of the queue. If a previous run hadn't finished before the next cron run, the size of the queue would increase and be processed down over time. `--cmd-run-frequency` does not inspect the queue and include its size in the calculation, but I think that's fine since a large queue of small objects (in terms of the size of their datastreams) will be processed much faster than a smaller queue of objects with very large datastreams (according to the results of my tinkering linked to above).

It might be a good idea to add an option to log a state where the queue had grown larger than a specified number of objects. But I would say that is beyond the scope of this PR. I can open a JIRA feature ticket for that.

**Comment by mjordan** [comment]
JIRA ticket for checking the size of the queue: https://jira.duraspace.org/browse/ISLANDORA-1921

**Comment by jonathangreen** [comment]
@mjordan that wording is great. I added a commit to update the README.

**Review by mjordan — APPROVED** [review]
@jonathangreen looks good, I'm merging this. Thanks!

File: islandora_checksum_checker.drush.inc
Comment (jordandukart) [inline islandora_checksum_checker.drush.inc:60]: This is a constant as opposed to `'required'`.

File: islandora_checksum_checker.drush.inc
Comment (jordandukart) [inline islandora_checksum_checker.drush.inc:65]: Same as https://github.com/Islandora/islandora_checksum_checker/pull/38/files#r111208451

Comment (DiegoPino) [reply islandora_checksum_checker.drush.inc:65]: @jordandukart i have a pull in some of my local branches for this. Today is a code heavy release day so will do it once i get out of work.

Comment (DiegoPino) [reply islandora_checksum_checker.drush.inc:65]: https://jira.duraspace.org/browse/ISLANDORA-1931?filter=13917

**Pull request merged**
